fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ExpressionProperty.java

```diff
@@ -20,158 +20,72 @@
- * the sort order and set of record types produced by a
- * {@link RelationalExpression} could be a
- * {@code ExpressionProperty}.
- *
- * <p>
- * To avoid littering {@link RelationalExpression} classes with methods for various properties, properties are implemented
- * using a variant of the hierarchical visitor pattern over a DAG of {@link RelationalExpression}s, {@link Quantifier}s,
- * and {@link Reference}s.
- * A property can be evaluated against an expression tree by having the visitor traverse a DAG of heterogeneous objects
- * where expressions are said to own quantifiers which range over expression references which then contain expressions
- * again. Shared subgraphs are visited multiple times. If desired, the caller must ensure that a subgraph is not
- * visited more than once.
- * </p>
- *
- * <p>
- * Note that the methods {@link #evaluateAtExpression}, {@link #evaluateAtQuantifier}, and {@link #evaluateAtRef} are handed the
- * results of the visitor evaluated at their owned quantifiers, references, and members respectively. Since most properties
- * are easy to describe as a recursion with depth one, this makes properties easier to read and write.
- * </p>
- *
- * <p>
- * Note that the default implementations provided in this interface walk the entire DAG, i.e. all {@link #shouldVisit}
- * methods return {@code true}. That implies that properties have been computed for sub-structures of the graph when
- * {@code evaluateAtXXX()} is called and therefore are not {@code null}. That means that all such arguments can be
- * assumed to be non-nullable even though parameters are annotated as {@link Nullable}. On the other hand, if the
- * implementor also overrides {@code shouldVisit} to return {@code false}, arguments to {@code evaluateAtXXX} may become
- * nullable or contain {@code null}s if they are collections and should be dealt with gracefully.
- * </p>
- *
- * @param <T> the result type of the property
+ * Base interface to capture properties for expressions.
+ * An instance of this usually class serves as a key in maps much like an enum, but provides strong typing.
+ * In particular, the following holds true for all properties:
+ * <ul>
+ *     <li>
+ *         Properties are singletons (or multitons; enumerated singletons) of implementors of this class.
+ *     </li>
+ *     <li>
+ *         Properties create a {@link RelationalExpressionVisitor} using the method {@link #createVisitor()}}. That
+ *         means that properties use visitors that operate on all kinds of relational expressions (from a Java typing
+ *         standpoint). In reality, as some properties are meaningless for some expressions but not others, the visitor
+ *         may decide to throw an exception for a class it cannot process.
+ *         There are some helpers that can transform e.g. a {@link RecordQueryPlanVisitor} into a
+ *         {@link RelationalExpressionVisitor} which add lowing to throw an exception when a non-plan is visited. This
```

Comment:
I think "lowing" here is probably a typo, but I'm not exactly sure what to replace it with. Perhaps:

```suggestion
 *         {@link RelationalExpressionVisitor} which will add throwing an exception when a non-plan is visited. This
```

Reply:
done.